### DISCUSSION TITLE
Inference for list comprehension and `functools.reduce`

### DISCUSSION THREAD

**Discussion by Viicos** [microsoft/pyright#7527] [Q&A] [answered]
Code sample in [pyright playground](https://pyright-play.net/?pyrightVersion=1.1.353&pythonVersion=3.12&pythonPlatform=Linux&strict=true&locale=en&code=JYWwDg9gTgLgBAMwK4DsDGMIQDYGcBQ%2BuMUuAXHNsMQNrFQC6cAvHDQEQCM7ANHOwCZe-AMzsGhZOkw5cAOigBTACZI0igBT44OygEMQAI2V64eznz0CK5uAGozAntt01gKGBsXYAlImhw3nDucPS4THAAxHAAglAA5kggih5wMACeYIrBuPxUtACqKADWKBAA7igM7Pg%2BhJEAtE1NhMqKCIioaBp6CRT5MG4eDD5kLjpgergE%2BFLdQ57efggBQSFhI0A)

```python
import functools

strs: list[str] = ["1", "2", "3"]

functools.reduce(
    lambda a1, a2: a1 + a2,
    [int(el) for el in strs]  # Argument type is "list[Unknown]"
)

#-----

def func(arg: list[int]):
    pass

func([int(el) for el in strs])
```

Asking as a question as this is maybe by design, but it seems `pyright` is not able to infer the type of `[int(el) for el in strs]` to be `list[int]` in the `functools.reduce` call. This felt weird for me as the second example with `func` seems to work fine (what I wanted to demonstrate is even without assigning the list to a temporary variable, the call to `func` still type checks).

**Comment by erictraut** [comment, marked as answer]
In both cases, pyright is applying [bidirectional type inference](https://microsoft.github.io/pyright/#/type-inference?id=bidirectional-type-inference-expected-types), so the inferred type of the comprehension is influenced by the "expected type". In the first example, the expected type is `Iterable[_T]`, and in the second example, the expected type is `Iterable[int]`. That makes the two cases quite different, as the first case relies on `_T` to be solved to influence the inferred type of the comprehension. 

Lambdas present a special challenge for type inference because their parameter types are never explicitly annotated, and their exact types are often difficult to infer.

I'll investigate further to see if there's a way I can improve the inference in this case.

**Reply by Viicos** [reply]
Ok this makes sense! If this ends up being to hard to support I don't think it is that much of a deal, as we can assign the new list to a variable first

**Reply by erictraut** [reply]
I tweaked the heuristics in the inference logic to handle this case. This will be included in the next release.

**Reply by Viicos** [reply]
Thanks for the ultra fast fix
